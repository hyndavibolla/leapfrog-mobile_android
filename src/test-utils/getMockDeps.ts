import { ENV } from '_constants';
import { GeneralModel, MissionModel } from '_models';
import { RUNTIME_ENV } from '_models/general';
import { ICrashlyticsService } from '_services/CrashlyticsService';
import { IHttpService } from '_services/HttpService';
import { IPerformanceTrackingService } from '_services/PerformanceTrackingService';
import { getInitialState } from '_state_mgmt/GlobalState';
import {
  getUser,
  getGame,
  getActivity_1,
  getActivity_2,
  getActivity_3,
  getMissionCategory_1,
  getMissionCategory_2,
  getMissionCategory_3,
  getMissionKeyword_1,
  getMissionKeyword_2,
  getMissionKeyword_3,
  getMission_1,
  getRewardConfig_1,
  getSlideObject_1,
  getSlideObject_2,
  getSlideCategoryIdList,
  getStreak_1,
  getQuestion_1,
  getQuestion_2,
  getQuestion_3,
  getAnswer_1,
  getAnswer_2,
  getAnswer_3,
  getNewOnMaxMessage_1,
  getNewOnMaxMessage_2,
  getNewOnMaxMessage_3,
  getSailthruMessage_1,
  getLinkedCards_1,
  getLinkedCards_2,
  getLinkedCards_3,
  getLinkedCards_4,
  getNewOnMaxMessage_4,
  getGiftCard,
  getGiftCardDetail,
  getGiftCardBalance
} from '_test_utils/entities';
import { createUUID } from '_utils/create-uuid';

export const getMockDeps = (initialStateGetter = getInitialState): GeneralModel.Deps => ({
  apiService: {
    isTokenInvalid: false,
    getDeviceIdAsync: async () => createUUID(),
    isTokenSetCurrentlyReady: () => true,
    setTokenSet: jest.fn(),
    authenticate: jest
      .fn()
      .mockResolvedValue({ accessToken: 'accessToken', refreshToken: 'refreshToken', accessTokenExpiryTime: 'accessTokenExpiryTime', deviceId: 'deviceId' }),
    enrollCard: jest.fn().mockResolvedValue(undefined),
    fetchRefId: jest.fn().mockResolvedValue('test-ref-id'),
    fetchUserProfile: jest.fn().mockResolvedValue(getUser()),
    fetchGameState: jest.fn().mockResolvedValue(getGame()),
    fetchRewardConfig: jest.fn().mockResolvedValue(getRewardConfig_1()),
    fetchActivityHistory: jest.fn().mockResolvedValue([getActivity_1(), getActivity_2(), getActivity_3()]),
    fetchGiftCardList: jest.fn().mockResolvedValue([getGiftCard()]),
    fetchGiftCardDetail: jest.fn().mockResolvedValue(getGiftCardDetail()),
    fetchGiftCardBalance: jest.fn().mockResolvedValue(getGiftCardBalance()),
    fetchWalletStats: jest.fn().mockResolvedValue({
      totalLifetimeEarnedPoints: 15550,
      monthlySummaries: {
        '2020-02': 1500,
        '2020-03': 2000,
        '2020-04': 100,
        '2020-05': 950,
        '2020-06': 1000,
        '2020-07': 1500
      }
    }),
    fetchMissionCategoryList: jest.fn().mockResolvedValue([getMissionCategory_1(), getMissionCategory_2(), getMissionCategory_3()]),
    fetchMissionKeywordList: jest.fn().mockResolvedValue({
      [MissionModel.KeywordType.CATEGORY]: [getMissionKeyword_1().value],
      [MissionModel.KeywordType.BRAND]: [getMissionKeyword_2().value, getMissionKeyword_3().value]
    }),
    fetchMissionList: jest.fn(async ({ listType }) => ({ userId: 'buttonId', missions: [getMission_1()], title: 'mission endpoint title', listType })),
    registerDeviceInfo: jest.fn(),
    registerUserActivity: jest.fn(),
    logout: jest.fn(),
    getSYWOnlineToken: jest.fn().mockResolvedValue('test-token'),
    sendValidationEmail: jest.fn().mockResolvedValue(undefined),
    fetchStreakList: jest.fn().mockResolvedValue([getStreak_1()]),
    fetchPromptList: jest.fn().mockResolvedValue({
      promptQuestions: [getQuestion_1(), getQuestion_2(), getQuestion_3()],
      answeredQuestions: [getAnswer_1(), getAnswer_2(), getAnswer_3()]
    }),
    updatePromptList: jest.fn(),
    unregisterDeviceInfo: jest.fn(),
    fetchLinkedCardsList: jest.fn().mockResolvedValue({ linkedCards: [getLinkedCards_1(), getLinkedCards_2(), getLinkedCards_3(), getLinkedCards_4()] }),
    activateLocalOffer: jest.fn(),
    fetchAppleWalletPass: jest.fn().mockResolvedValue('test-pass'),
    updateCardStatus: jest.fn()
  } as any,
  appService: {
    getDeviceIdAsync: async () => createUUID()
  } as any,
  crashlyticsService: {
    initialize: jest.fn(),
    log: jest.fn(),
    recordError: jest.fn(),
    setCrashlyticsCollectionEnabled: jest.fn()
  } as ICrashlyticsService,
  httpService: {
    fetch: jest.fn()
  } as IHttpService,
  performanceTrackingService: {
    initialize: jest.fn(),
    measuredFetch: jest.fn(),
    setPerformanceCollectionEnabled: jest.fn()
  } as IPerformanceTrackingService,
  raiseService: {
    fetchSlideBrand: jest.fn().mockResolvedValue([getSlideObject_1(), getSlideObject_2()]),
    fetchSlideCategoryIdList: jest.fn().mockResolvedValue(getSlideCategoryIdList())
  } as any,
  nativeHelperService: {
    alert: {
      alert: jest.fn(),
      prompt: jest.fn()
    },
    appState: { addEventListener: jest.fn(), currentState: 'active' },
    storage: {
      get: jest.fn().mockResolvedValue(null),
      set: jest.fn().mockResolvedValue(null),
      remove: jest.fn().mockResolvedValue(null)
    },
    clipboard: {
      getString: jest.fn(),
      setString: jest.fn()
    },
    dimensions: {
      getWindowWidth: jest.fn().mockReturnValue(750),
      getWindowHeight: jest.fn().mockReturnValue(1000),
      isSmallDevice: false
    },
    deviceInfo: {
      getDeviceId: createUUID,
      getUniqueId: createUUID,
      getVersion: () => '0.0',
      getBuildNumber: () => '0.1',
      getDeviceModel: () => 'iPhone 11',
      getLocales: { languageTag: 'en-US' },
      getBaseOs: async () => 'iOS',
      getSystemVersion: () => '11.0',
      getDeviceToken: () => createUUID(),
      getDeviceTokenAsync: async () => createUUID(),
      getDeviceOsType: async () => 'iOS',
      getAppVersion: () => '0.0.0.1',
      getDeviceLanguage: () => 'en-US',
      getDeviceOsVersion: () => '11.0',
      getFCMTokenAsync: async () => createUUID(),
      getTimeZone: () => 'testing-timezone',
      getReadableVersion: () => '1.1.1.1',
      onFCMTokenRefresh: jest.fn()
    },
    netInfo: {
      getState: jest.fn().mockResolvedValue({ isConnected: true }),
      subscribeToConnectionChanges: jest.fn(),
      waitForConnection: jest.fn().mockResolvedValue(undefined)
    },
    linking: {
      openURL: jest.fn(() => Promise.resolve()),
      canOpenURL: jest.fn(),
      checkURLScheme: jest.fn().mockResolvedValue(true),
      getInitialURL: jest.fn(),
      addEventListener: jest.fn()
    },
    buttonSdk: {
      configure: jest.fn(),
      setUserId: jest.fn(),
      clearData: jest.fn(),
      purchaseRequest: jest.fn()
    },
    platform: {
      os: 'default',
      select: jest.fn(config => config?.default)
    },
    fireBaseMessaging: {
      getToken: jest.fn().mockReturnValue(createUUID())
    },
    hash: {
      buildSHA256: jest.fn().mockResolvedValue('hash')
    },
    restart: jest.fn(),
    cookieManager: {
      clearAll: jest.fn()
    },
    base64: {
      encode: jest.fn((...args) => `b64 encoded: ${JSON.stringify(args)}`),
      decode: jest.fn((...args) => `b64 decoded: ${JSON.stringify(args)}`)
    },
    getRuntimeEnv: jest.fn().mockResolvedValue(RUNTIME_ENV.DEV_TEST),
    messaging: {
      hasPermission: jest.fn().mockReturnValue(true),
      requestPermission: jest.fn()
    },
    pushNotificationsAuthStatus: {
      NOT_DETERMINED: -1,
      DENIED: 0,
      AUTHORIZED: 1,
      PROVISIONAL: 2
    },
    sailthru: {
      getMessages: jest
        .fn()
        .mockResolvedValue([getNewOnMaxMessage_1(), getNewOnMaxMessage_4(), getNewOnMaxMessage_2(), getNewOnMaxMessage_3(), getSailthruMessage_1()]),
      markMessageAsRead: jest.fn().mockResolvedValue(true),
      getInitialNotification: jest.fn(),
      onNotificationTapped: jest.fn(cb => {
        cb({});
        return () => {};
      }) as any
    },
    reactNativePermission: require('react-native-permissions/mock'),
    geolocation: {
      addListener: jest.fn(),
      getCurrentPosition: jest.fn(),
      removeListeners: jest.fn(),
      requestAuthorization: jest.fn(),
      setConfiguration: jest.fn(),
      startObserving: jest.fn(),
      stopObserving: jest.fn()
    }
  } as any,
  stateSnapshot: {
    set: jest.fn(),
    get: jest.fn().mockReturnValue(initialStateGetter())
  },
  logger: {
    log: jest.fn(),
    info: jest.fn(),
    debug: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
    success: jest.fn(),
    color: jest.fn(),
    assert: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    format: jest.fn()
  } as any,
  eventTrackerService: {
    trackDataUser: jest.fn(),
    isTrackingEnabled: jest.fn().mockResolvedValue(true),
    rZero: {
      trackView: jest.fn(),
      trackEvent: jest.fn(),
      flush: jest.fn(),
      setUser: jest.fn()
    },
    forterSDK: {
      setDevLogsEnabled: jest.fn(),
      getDeviceUniqueID: jest.fn(cb => cb()),
      init: jest.fn((a, b, c) => c()),
      trackNavigation: jest.fn(),
      trackNavigationWithExtraData: jest.fn(),
      trackAction: jest.fn(),
      trackActionWithJSON: jest.fn(),
      fraudTrackNavigationEvent: jest.fn(),
      fraudTrackAction: jest.fn()
    },
    tealiumSDK: {
      track: jest.fn(),
      trackView: jest.fn()
    },
    appsFlyerSDK: {
      updateServerUninstallToken: jest.fn(),
      setCustomerUserId: jest.fn().mockImplementation((_, cb) => cb('success')),
      addDeepLinkEventListener: jest.fn(),
      removeDeepLinkEventListener: jest.fn(),
      clearDeepLinkEventListener: jest.fn(),
      trackConversionEvent: jest.fn()
    }
  } as any,
  ccpaService: {
    showConsentBanner: jest.fn().mockResolvedValue(null),
    showPreferenceCenter: jest.fn()
  } as any,
  remoteConfigService: {
    getValue: jest.fn(async key => ENV.REMOTE_CONFIG.DEFAULT_CONFIG[key]),
    fetchAndActivate: jest.fn().mockResolvedValue(null)
  } as any,
  authService: {
    auth: {
      refreshToken: jest.fn().mockResolvedValue(true)
    },
    webAuth: {
      authorize: jest.fn(),
      clearSession: jest.fn()
    }
  } as any
});
