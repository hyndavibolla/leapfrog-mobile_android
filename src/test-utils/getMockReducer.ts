import { IGlobalState } from '../models/general';
import { combinedReducer } from '../state-mgmt/GlobalState';

export const getMockReducer = (state?: IGlobalState) => jest.fn((s, a) => state || combinedReducer(s, a));
