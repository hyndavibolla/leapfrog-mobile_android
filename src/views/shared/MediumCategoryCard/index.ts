export { default as MediumCategoryCard } from './MediumCategoryCard';
