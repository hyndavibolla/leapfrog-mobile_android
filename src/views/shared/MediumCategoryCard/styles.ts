import { css } from '@emotion/native';

import { COLOR, colorWithOpacity, FONT_FAMILY, FONT_SIZE } from '../../../constants';

export const styles = {
  defaultBoxStyle: css`
    border-radius: 8px;
    width: 100%;
    height: 100px;
  `,
  backgroundContainer: css`
    background-color: ${COLOR.PRIMARY_BLUE};
    overflow: hidden;
  `,
  container: css`
    flex: 1;
    align-items: stretch;
    justify-content: center;
    background-color: ${colorWithOpacity(COLOR.BLACK, 45)};
  `,
  title: css`
    color: ${COLOR.WHITE};
    font-size: ${FONT_SIZE.SMALL};
    font-family: ${FONT_FAMILY.HEAVY};
    text-align: center;
    padding-horizontal: 20px;
  `
};
