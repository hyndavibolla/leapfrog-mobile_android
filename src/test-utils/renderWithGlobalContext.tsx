import React, { ReactElement } from 'react';
import { render } from '@testing-library/react-native';

import { getInitialState, GlobalProvider } from '../state-mgmt/GlobalState';
import { Deps, IGlobalState, IReducer } from '../models/general';
import { getMockDeps } from './getMockDeps';
import { getMockReducer } from './getMockReducer';

export const renderWithGlobalContext = (
  children: ReactElement,
  deps: Deps = getMockDeps(),
  initState: IGlobalState = getInitialState(),
  combinedReducers: IReducer<IGlobalState, any> = getMockReducer()
) => {
  const renderAPI = render(
    <GlobalProvider deps={deps} initState={initState} combinedReducers={combinedReducers}>
      {children}
    </GlobalProvider>
  );
  return { ...renderAPI, mockReducer: combinedReducers, deps };
};
