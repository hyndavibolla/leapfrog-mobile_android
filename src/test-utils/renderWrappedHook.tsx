import React from 'react';
import { renderHook } from '@testing-library/react-hooks';

import { getInitialState, GlobalProvider } from '../state-mgmt/GlobalState';
import { Deps, IGlobalState, IReducer } from '../models/general';
import { getMockDeps } from './getMockDeps';
import { getMockReducer } from './getMockReducer';

export const renderWrappedHook = (
  hookGetter: () => any,
  deps: Deps = getMockDeps(),
  initState: IGlobalState = getInitialState(),
  combinedReducers: IReducer<IGlobalState, any> = getMockReducer()
) => {
  const wrapper = ({ children }: any) => (
    <GlobalProvider deps={deps} initState={initState} combinedReducers={combinedReducers}>
      {children}
    </GlobalProvider>
  );
  return { ...renderHook(() => hookGetter(), { wrapper }), mockReducer: combinedReducers, deps };
};
