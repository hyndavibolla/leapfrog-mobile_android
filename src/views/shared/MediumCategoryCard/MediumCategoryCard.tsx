import React, { memo, useMemo } from 'react';
import { TouchableHighlight, View } from 'react-native';
import { ReactNativeStyle } from '@emotion/native';

import { CONTAINER_STYLE } from '_constants';
import { ImageBackgroundWithFallback } from '_components/ImageBackgroundWithFallback';
import { Text } from '_components/Text';
import { useTestingHelper } from '_utils/useTestingHelper';

import { styles } from './styles';

export interface Props {
  title: string;
  backgroundUrl?: string;
  backgroundColor?: string;
  boxStyle?: ReactNativeStyle;
  onPress?: () => void;
  itemContainerStyle?: ReactNativeStyle;
  textStyle?: ReactNativeStyle;
}

export const MediumCategoryCard = ({ title, backgroundUrl, boxStyle, onPress, itemContainerStyle, textStyle }: Props) => {
  const { getTestIdProps } = useTestingHelper('medium-category-card');
  boxStyle = { ...styles.defaultBoxStyle, ...boxStyle };

  const containerStyle = useMemo(() => [CONTAINER_STYLE.shadow, styles.backgroundContainer, boxStyle], [boxStyle]);

  return (
    <TouchableHighlight underlayColor="transparent" onPress={onPress} {...getTestIdProps('container')}>
      {backgroundUrl ? (
        <ImageBackgroundWithFallback
          containerStyle={containerStyle}
          source={{ uri: backgroundUrl }}
          fallbackSource={require('_assets/shared/categoryBackgroundFallback.png')}
        >
          <View style={[styles.container, itemContainerStyle]}>
            <Text style={[styles.title, textStyle]} numberOfLines={2} ellipsizeMode="tail">
              {title}
            </Text>
          </View>
        </ImageBackgroundWithFallback>
      ) : (
        <View style={containerStyle}>
          <View style={[styles.container, itemContainerStyle]}>
            <Text style={[styles.title, textStyle]} numberOfLines={2} ellipsizeMode="tail">
              {title}
            </Text>
          </View>
        </View>
      )}
    </TouchableHighlight>
  );
};

export default memo(MediumCategoryCard);
