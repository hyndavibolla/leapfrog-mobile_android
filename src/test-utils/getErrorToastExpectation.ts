import { act } from 'react-test-renderer';

import { ACTION_TYPE } from '../state-mgmt/core/actions';
import { ToastType } from '../views/shared/Toast';
import { Deps } from '../models/general';
import { wait } from '../utils/wait';
import { renderWrappedHook } from './renderWrappedHook';

export const getErrorToastExpectation = async (hookGetter: () => any, deps: Deps, errorArgs: any[] = []) => {
  const { result, mockReducer } = renderWrappedHook(hookGetter, deps);
  await act(async () => {
    await (result.current[0] as (...args: any[]) => Promise<any>)(...errorArgs);
    await wait(0); // waiting for the useEffect dispatching the toast to run
    expect(mockReducer).toBeCalledWith(
      expect.any(Object),
      expect.objectContaining({
        type: ACTION_TYPE.ADD_TOAST,
        payload: expect.objectContaining({ props: expect.objectContaining({ type: ToastType.ERROR }) })
      })
    );
  });
};
