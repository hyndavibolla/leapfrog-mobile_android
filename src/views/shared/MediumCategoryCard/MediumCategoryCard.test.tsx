import React from 'react';
import { fireEvent } from '@testing-library/react-native';

import MediumCategoryCard, { Props } from './MediumCategoryCard';
import { getMissionCategory_1 } from '../../../test-utils/entities';
import { renderWithGlobalContext } from '../../../test-utils/renderWithGlobalContext';

describe('MediumCategoryCard', () => {
  let props: Props;
  beforeEach(() => {
    const category = getMissionCategory_1();
    props = {
      title: category.name,
      backgroundUrl: category.lifestyleUrl,
      onPress: jest.fn()
    };
  });

  it('should render', () => {
    const { toJSON } = renderWithGlobalContext(<MediumCategoryCard {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should be pressable', () => {
    const { getByTestId } = renderWithGlobalContext(<MediumCategoryCard {...props} />);
    fireEvent.press(getByTestId('medium-category-card-container'));
    expect(props.onPress).toBeCalled();
  });
});
