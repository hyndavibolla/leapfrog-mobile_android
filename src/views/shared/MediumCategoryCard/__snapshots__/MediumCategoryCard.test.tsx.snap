// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediumCategoryCard should render 1`] = `
<View
  accessibilityLabel="medium-category-card-container"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="medium-category-card-container"
>
  <View
    accessibilityLabel="image-background-container"
    accessible={true}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "elevation": 3,
          "shadowColor": "#262C2F",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.1,
        },
        Object {
          "backgroundColor": "#0066CC",
          "overflow": "hidden",
        },
        Object {
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
          "height": 100,
          "width": "100%",
        },
      ]
    }
    testID="image-background-container"
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": 100,
              "width": "100%",
            },
            undefined,
          ],
        ]
      }
    >
      <FastImageView
        accessibilityLabel="image-background-image"
        accessible={true}
        onFastImageError={[Function]}
        onFastImageLoadEnd={[Function]}
        onFastImageLoadStart={[Function]}
        resizeMode="cover"
        source={
          Object {
            "uri": "https://i.sears.com/s/i/or/SYW-MAX/Brands_Apparel_&_Accessories-_AGGGFVI9CG",
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        testID="image-background-image"
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "stretch",
              "backgroundColor": "#262C2F45",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={2}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "SFProDisplay-Heavy",
                  "fontSize": 18,
                  "paddingHorizontal": 20,
                  "textAlign": "center",
                },
                undefined,
              ],
            ]
          }
        >
          Apparel & Accessories
        </Text>
      </View>
    </View>
  </View>
</View>
`;
